## Project: Profitable App Profiles for the App Store and Google Play Markets

In this project, our goal is to analyze data to assist developers understand what type of apps are likley to attact more users on Google Play and the App Store. 

As of September 2018, there were approximately 2 million iOS apps available on the App Store, and 2.1 million Android apps on Google Play.

Collecting data for over 4 million apps requires a significant amount of time and money, so we'll try to analyze a sample of the data instead. To avoid spending resources on collecting new data ourselves, we should first try to see if we can find any relevant existing data at no cost. Luckily, these are two data sets that seem suitable for our goals:

- [A data set](https://www.kaggle.com/lava18/google-play-store-apps) containing data about approximately 10,000 Android apps from Google Play; the data was collected in August 2018. You can download the data set directly from [this link.](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv)

- [A data set](https://www.kaggle.com/ramamet4/app-store-apple-data-set-10k-apps) containing data about approximately 7,000 iOS apps from the App Store; the data was collected in July 2017. You can download the data set directly from [this link.](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)


In [26]:
from csv import reader

#Google#
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]
len_android_columns = 'len_android_columns = '+str(len(android_header))
len_android_rows = 'len_android_rows = '+str(len(android))

#Apple#
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]
len_ios_columns = 'len_ios_columns = '+str(len(ios_header))
len_ios_rows = 'len_ios_rows = ' +str(len(ios))

To make it easier to explore the two data sets, we'll first write a function named explore_data() that we can use repeatedly to explore rows in a more readable way. We'll also add an option for our function to show the number of rows and columns for any data set.

In [27]:
#explore data function - provided#

def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]
    for row in dataset_slice:
        print(row,'\n')
        #adds a new empty line after each row
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns', len(dataset[0]))

In [28]:
#google exploring#
print('~ GOOGLE DATA REVIEW ~')
print('\n')
print(android_header)
print('\n')
explore_data(android, 0, 3, True)
print('\n')
print('\n')

#apple exploring#
print('~ IOS DATA REVIEW ~')
print('\n')
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

~ GOOGLE DATA REVIEW ~


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

Number of rows: 10841
Number of columns 13




~ IOS DATA REVIEW ~


['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipad

In [29]:
print(android[10473])
print('\n')
print(android_header)
print('\n')
print(android[0])

['osmino Wi-Fi: free WiFi', 'TOOLS', '4.2', '134203', '4.1M', '10,000,000+', 'Free', '0', 'Everyone', 'Tools', 'August 7, 2018', '6.06.14', '4.4 and up']


['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


## Deleting Wrong Data

The Google Play data set has a [dedicated discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, and we can see that [one of the discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion/164101) outlines an error for row 10472. Let's print this row and compare it against the header and another row that is correct.

In [30]:
print(len(android))
del android[10472]
print(len(android))


10841
10840


## Removing Duplicate Entries

**Part One**

Upon further review of the Google Play data and [discussions](https://www.kaggle.com/lava18/google-play-store-apps/discussion) section, we found that some apps have more than one entry. For instance, the application Instagram has several duplicate entries:

In [31]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


## Identifiying Number of Duplicate Apps in GooglePlay Dataset

We will create two lists, one for duplicate apps and one for unique apps contained in the GooglePlay dataset. This is completed by for loop through the android dataset and for each iteration we will do the following:

- Save the app name to a variable named *name*
- If *name* was already in the unique apps list, we append that *name* to the duplicate apps list. 
- Else we will append the *name* into the unique apps list. 

Lastly, we will print the length of each (*duplicate names*, and *unique names*) to gain an understanding of how many apps need to be removed from the dataset. 

In [32]:
duplicate_apps = []
unique_apps = []

for app in android:
    name = app[0]
    if name in unique_apps:
        duplicate_apps.append(name)
    else:
        unique_apps.append(name)
        
print('Number of duplicate apps:', len(duplicate_apps))
print('\n')    
print('Examples of duplicate apps:', duplicate_apps[:15])
print('\n')
print('Number of unique apps:', len(unique_apps))

Number of duplicate apps: 1181


Examples of duplicate apps: ['Quick PDF Scanner + OCR FREE', 'Box', 'Google My Business', 'ZOOM Cloud Meetings', 'join.me - Simple Meetings', 'Box', 'Zenefits', 'Google Ads', 'Google My Business', 'Slack', 'FreshBooks Classic', 'Insightly CRM', 'QuickBooks Accounting: Invoicing & Expenses', 'HipChat - Chat Built for Teams', 'Xero Accounting Software']


Number of unique apps: 9659


## Removing Duplicate Apps in GooglePlay Dataset

We now understand there are 1,181 duplicate apps in the andriod dataset. After we remove the duplicates, we will expect to be left with 9,659 rows of unique apps. 

To remove duplicates we will do the following:
- Create a dictionary where each dictionary key is a unique app name and the corresponding value is the highest number of reviews of that app. (we have chosen this value because it is unique for each row, even for duplicate apps in the andriod data set, and we will only want to keep the row with the *highest* reviews of the app as it will be the most updated version of that information)
- Use the information stored in the dictionary in creation of a new dataset, which will have only one entry per app (again, only selected with the *highest* number for reviews)

In [33]:
reviews_max = {}

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    
    elif name not in reviews_max:
        reviews_max[name] = n_reviews

print('Length of Reviews_Max Dictionary:',len(reviews_max))

android_clean = []
already_added = []

for app in android:
    name = app[0]
    n_reviews = float(app[3])
    
    if (reviews_max[name] == n_reviews) and (name not in already_added):
        android_clean.append(app)
        already_added.append(name)

print('Length of Android_Cleaned List:',len(android_clean))
print('\n')
explore_data(android_clean, 0, 3, True)

Length of Reviews_Max Dictionary: 9659
Length of Android_Cleaned List: 9659


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows: 9659
Number of columns 13


## Removing Non-English Apps from App Store Dataset

Previously, we manged to remove duplicate app entries in the GooglePlay dataset. The App Store does not contain duplciates, but does contain non-english speaking apps. We will remove the non-engish speaking apps from the App Store dataset by completing the following:

**NOTE: Since we understand each character used in a string has a corresponding number associated with it, and accoridng to the [ASCII](https://en.wikipedia.org/wiki/ASCII)(American Standard Code for Information Interchange) system, english characters are in the range 0 to 127. Therefore, any characters that are equal to or less than 127 then we know the character belongs to the English language.**

- Write a function that takes in a string and returns *FALSE* if there is any character in the string that does not belong to the set of English characters (<=127)
 - In the function, iterate over the input string and for each iteration check whether the number is associated with a character > 127 and if so, should return *FALSE*
 - If the loop finished running without the retunr statement being excuted, then it means no characters had a corresponding number > 127 & the app name is English and should return *TRUE*. 
- We will test the following names to confirm if they are detected as English or Non-English:
 - 'Instagram'
 - '爱奇艺PPS -《欢乐颂2》电视剧热播'
 - 'Docs To Go™ Free Office Suite'
 - 'Instachat 😜'


In [34]:
def is_english(string):
    for character in string:
        if ord(character) > 127:
            return False
    
    return True

#Testing if function works
print(is_english('Instagram'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('Instachat 😜'))


True
False
False
False


In reviewing the results of the function detecting non-english apps, we can see that if there are special characters in the English app name, it will return *FALSE*. Emojis and characters like ™ fall outside the ASCII range and have corresponding numbers greater than 127. 

See results below for corresponding numbers to 😜 and ™. 

In [35]:
print(ord('™'))
print(ord('😜'))

8482
128540


To minimize loosing useful data, we will only remove an app if the name has more than three characters with corresponding numbers greater than 127. 


In [36]:
def is_english(string):
    non_ascii = 0
    
    for character in string:
        if ord(character) > 127:
            non_ascii += 1
    
    if non_ascii > 3:
        return False
    else:
        return True

print(is_english('Instachat 😜'))   
print(is_english('Docs To Go™ Free Office Suite'))
print(is_english('爱奇艺PPS -《欢乐颂2》电视剧热播'))


True
True
False


Below, we use the is_english() function to filter out the non-English apps for both data sets;

In [39]:
android_english = []
ios_english = []

for app in android_clean:
    name = app[0]
    if is_english(name):
        android_english.append(app)
        
for app in ios:
    name = app[1]
    if is_english(name):
        ios_english.append(app)
        
explore_data(android_english, 0, 3, True)
print('\n')
explore_data(ios_english, 0, 3, True)

['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up'] 

['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up'] 

['Sketch - Draw & Paint', 'ART_AND_DESIGN', '4.5', '215644', '25M', '50,000,000+', 'Free', '0', 'Teen', 'Art & Design', 'June 8, 2018', 'Varies with device', '4.2 and up'] 

Number of rows: 9614
Number of columns 13


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1'] 

['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1'] 

['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', '

From the above, we can see that we're left with **9,614** Android apps and **6,183** iOS apps; **1,5797** in total.

To recap, during the data cleaning process we have completed the following:
- Removed inaccurate data
- Removed duplicate app entries
- Removed non-english apps

Next we will be isolating the free apps as our last step in the data cleaning process. 

## Isolating free apps from GooglePlay and IOS Dataset

We will loop through each data set to isolate the free apps in seperate lists & careful to identify the correct column describing the app price. 

After we isolate the free apps in two lists, we will check the lenght of each dataset to see how many apps are remaining.

In [42]:
ios_final = []
android_final = []

for app in android_english:
    price = app[7]
    if price == '0':
        android_final.append(app)
        
for app in ios_english:
    price = app[4]
    if price == '0.0':
        ios_final.append(app)

print(len(android_final))        
print(len(ios_final))

8864
3222


We are left with **8,864** Android apps and **3,222** iOS apps; **12,086** in total that are free english apps. 

## Most Common Apps by Genre

**Part ONE**

So far we have concluded our data cleaning process with the following: 
- Removed inaccurate data
- Removed duplicate app entries
- Removed non-english apps
- Isolated free apps on the Android and IOS markets

Our aim is to determin the kinds of apps that are likely to attract more uses due to revenue being highly influenced by the number of people using the apps. 

In aim to minimize risks and overhead, our validation strategy for an app idea is comprised of the following steps:

1. Build a minimal Android verson of the app & add it to GooglePlay
2. If the app obtains good response from users, it will be developed further. 
3. if the app if profitable after ~6 months, we will build IOS verson of the app & add to App Store. 

We will need to find app profiles that are successful on both markets. For start, we will get a sense of what are the most common geners for each market by building frequeny tables for a few columns in the datasets. 

In [48]:
print('~ Google Final Headers ~')
print(android_final[0])
print('\n')
print('~ IOS Final Headers ~')
print(ios_final[0])

~ Google Final Headers ~
['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


~ IOS Final Headers ~
['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


**Part TWO**

We'll build two functions we can use to analyze the frequency tables:

One function to generate frequency tables that show percentages
Another function that we can use to display the percentages in a descending order



In [50]:
def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1 
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
        
    table_percentages = {}
    for key in table:
        percentage = (table[key] / total) * 100
        table_percentages[key] = percentage
 
    
    return table_percentages

def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])
        

**Part THREE**

We start by examining the frequency table for the prime_genre column of the App Store data set.

In [57]:
print('~ GOOGLEPLAY CATEGORYS ~')
display_table(android_final, 1)

~ GOOGLEPLAY CATEGORYS ~
FAMILY : 18.907942238267147
GAME : 9.724729241877256
TOOLS : 8.461191335740072
BUSINESS : 4.591606498194946
LIFESTYLE : 3.9034296028880866
PRODUCTIVITY : 3.892148014440433
FINANCE : 3.7003610108303246
MEDICAL : 3.531137184115524
SPORTS : 3.395758122743682
PERSONALIZATION : 3.3167870036101084
COMMUNICATION : 3.2378158844765346
HEALTH_AND_FITNESS : 3.0798736462093865
PHOTOGRAPHY : 2.944494584837545
NEWS_AND_MAGAZINES : 2.7978339350180503
SOCIAL : 2.6624548736462095
TRAVEL_AND_LOCAL : 2.33528880866426
SHOPPING : 2.2450361010830324
BOOKS_AND_REFERENCE : 2.1435018050541514
DATING : 1.861462093862816
VIDEO_PLAYERS : 1.7937725631768955
MAPS_AND_NAVIGATION : 1.3989169675090252
FOOD_AND_DRINK : 1.2409747292418771
EDUCATION : 1.1620036101083033
ENTERTAINMENT : 0.9589350180505415
LIBRARIES_AND_DEMO : 0.9363718411552346
AUTO_AND_VEHICLES : 0.9250902527075812
HOUSE_AND_HOME : 0.8235559566787004
WEATHER : 0.8009927797833934
EVENTS : 0.7107400722021661
PARENTING : 0.654332129

In [60]:
print('~ GOOGLEPLAY GENRES ~')
display_table(android_final, -4)

~ GOOGLEPLAY GENRES ~
Tools : 8.449909747292418
Entertainment : 6.069494584837545
Education : 5.347472924187725
Business : 4.591606498194946
Productivity : 3.892148014440433
Lifestyle : 3.892148014440433
Finance : 3.7003610108303246
Medical : 3.531137184115524
Sports : 3.463447653429603
Personalization : 3.3167870036101084
Communication : 3.2378158844765346
Action : 3.1024368231046933
Health & Fitness : 3.0798736462093865
Photography : 2.944494584837545
News & Magazines : 2.7978339350180503
Social : 2.6624548736462095
Travel & Local : 2.3240072202166067
Shopping : 2.2450361010830324
Books & Reference : 2.1435018050541514
Simulation : 2.0419675090252705
Dating : 1.861462093862816
Arcade : 1.8501805054151623
Video Players & Editors : 1.7712093862815883
Casual : 1.7599277978339352
Maps & Navigation : 1.3989169675090252
Food & Drink : 1.2409747292418771
Puzzle : 1.128158844765343
Racing : 0.9927797833935018
Role Playing : 0.9363718411552346
Libraries & Demo : 0.9363718411552346
Auto & Vehi

In [63]:
print('~ IOS GENRES ~')
display_table(ios_final, -5)

~ IOS GENRES ~
Games : 58.16263190564867
Entertainment : 7.883302296710118
Photo & Video : 4.9658597144630665
Education : 3.662321539416512
Social Networking : 3.2898820608317814
Shopping : 2.60707635009311
Utilities : 2.5139664804469275
Sports : 2.1415270018621975
Music : 2.0484171322160147
Health & Fitness : 2.0173805090006205
Productivity : 1.7380509000620732
Lifestyle : 1.5828677839851024
News : 1.3345747982619491
Travel : 1.2414649286157666
Finance : 1.1173184357541899
Weather : 0.8690254500310366
Food & Drink : 0.8069522036002483
Reference : 0.5586592178770949
Business : 0.5276225946617008
Book : 0.4345127250155183
Navigation : 0.186219739292365
Medical : 0.186219739292365
Catalogs : 0.12414649286157665


## Review of most Common Apps by Genre

The frequeny table above shows the App store is dominated by apps desgined for fun *(Games 58%)*, while GooglePlay shows a more balanced landscape of both practical and fun apps. We would like to get an idea about the kind of apps with the most users. 

To find which genres have the most users, we will calculate the average number of installs for each app genre. 
 - GooglePlay we will use information from the Installs column.
 - This information is missing from the App Store dataset and as a result, we will create a workaround by taking the total number of user ratings as a proxy, which can be found in the rating_count_tot app. 
 
 We will start with calculating the average number of user ratings per app genre on the App Store;
 - Isolate the apps of genre
 - Sum up the user ratings for the apps of that genre
 - Divide the sum by the number of apps belonging to that genre ( **not** by the total number of apps)

To calculate the avg number of user ratings for each genre, we will use a for loop inside of another for loop (*nested loop*).  
 


## Most Popular Apps by Genre on the App Store

In [81]:
genres_ios = freq_table(ios_final, -5)

for genre in genres_ios:
    total = 0
    len_genre = 0
    for app in ios_final:
        genre_app = app[-5]
        if genre_app == genre:            
            n_ratings = float(app[5])
            total += n_ratings
            len_genre += 1
    avg_n_ratings = total / len_genre
    print(genre, ':', avg_n_ratings)

Shopping : 26919.690476190477
Games : 22788.6696905016
Entertainment : 14029.830708661417
Food & Drink : 33333.92307692308
Sports : 23008.898550724636
Navigation : 86090.33333333333
News : 21248.023255813954
Catalogs : 4004.0
Health & Fitness : 23298.015384615384
Book : 39758.5
Reference : 74942.11111111111
Finance : 31467.944444444445
Medical : 612.0
Utilities : 18684.456790123455
Social Networking : 71548.34905660378
Music : 57326.530303030304
Productivity : 21028.410714285714
Education : 7003.983050847458
Travel : 28243.8
Weather : 52279.892857142855
Business : 7491.117647058823
Photo & Video : 28441.54375
Lifestyle : 16485.764705882353


** IOS App Recommendations:** 

(>= 70,000 avg ratings per genre)
- Navigation
- Weather
- Social Networking
- Music
- Reference

We will drill down into the genres to understand what is contributing to the high user ratings for each.

In [91]:
print('~ IOS NAVIGATION ~')

for app in ios_final:
    if app[-5] == 'Navigation':
        print(app[1], ':', app[5]) # print name and number of ratings

~ IOS NAVIGATION ~
Waze - GPS Navigation, Maps & Real-time Traffic : 345046
Google Maps - Navigation & Transit : 154911
Geocaching® : 12811
CoPilot GPS – Car Navigation & Offline Maps : 3582
ImmobilienScout24: Real Estate Search in Germany : 187
Railway Route Search : 5


In [85]:
print('~ IOS Weather ~')
for app in ios_final:
    if app[-5] == 'Weather':
        print(app[1], ':', app[5])

~ IOS Weather ~
The Weather Channel: Forecast, Radar & Alerts : 495626
The Weather Channel App for iPad – best local forecast, radar map, and storm tracking : 208648
WeatherBug - Local Weather, Radar, Maps, Alerts : 188583
MyRadar NOAA Weather Radar Forecast : 150158
AccuWeather - Weather for Life : 144214
Yahoo Weather : 112603
Weather Underground: Custom Forecast & Local Radar : 49192
NOAA Weather Radar - Weather Forecast & HD Radar : 45696
Weather Live Free - Weather Forecast & Alerts : 35702
Storm Radar : 22792
QuakeFeed Earthquake Map, Alerts, and News : 6081
Moji Weather - Free Weather Forecast : 2333
Hurricane by American Red Cross : 1158
Forecast Bar : 375
Hurricane Tracker WESH 2 Orlando, Central Florida : 203
FEMA : 128
iWeather - World weather forecast : 80
Weather - Radar - Storm with Morecast App : 78
Yurekuru Call : 53
Weather & Radar : 37
WRAL Weather Alert : 25
Météo-France : 24
JaxReady : 22
Freddy the Frogcaster's Weather Station : 14
Almanac Long-Range Weather Foreca

In [88]:
print('~ IOS SOCIAL NETWORKING ~')

for app in ios_final:
    if app[-5] == 'Social Networking':
        print(app[1], ':', app[5])

~ IOS SOCIAL NETWORKING ~
Facebook : 2974676
Pinterest : 1061624
Skype for iPhone : 373519
Messenger : 351466
Tumblr : 334293
WhatsApp Messenger : 287589
Kik : 260965
ooVoo – Free Video Call, Text and Voice : 177501
TextNow - Unlimited Text + Calls : 164963
Viber Messenger – Text & Call : 164249
Followers - Social Analytics For Instagram : 112778
MeetMe - Chat and Meet New People : 97072
We Heart It - Fashion, wallpapers, quotes, tattoos : 90414
InsTrack for Instagram - Analytics Plus More : 85535
Tango - Free Video Call, Voice and Chat : 75412
LinkedIn : 71856
Match™ - #1 Dating App. : 60659
Skype for iPad : 60163
POF - Best Dating App for Conversations : 52642
Timehop : 49510
Find My Family, Friends & iPhone - Life360 Locator : 43877
Whisper - Share, Express, Meet : 39819
Hangouts : 36404
LINE PLAY - Your Avatar World : 34677
WeChat : 34584
Badoo - Meet New People, Chat, Socialize. : 34428
Followers + for Instagram - Follower Analytics : 28633
GroupMe : 28260
Marco Polo Video Walkie 

In [89]:
print('~ Music ~')

for app in ios_final:
    if app[-5] == 'Music': 
        print(app[1], ':', app[5])

~ Music ~
Pandora - Music & Radio : 1126879
Spotify Music : 878563
Shazam - Discover music, artists, videos & lyrics : 402925
iHeartRadio – Free Music & Radio Stations : 293228
SoundCloud - Music & Audio : 135744
Magic Piano by Smule : 131695
Smule Sing! : 119316
TuneIn Radio - MLB NBA Audiobooks Podcasts Music : 110420
Amazon Music : 106235
SoundHound Song Search & Music Player : 82602
Sonos Controller : 48905
Bandsintown Concerts : 30845
Karaoke - Sing Karaoke, Unlimited Songs! : 28606
My Mixtapez Music : 26286
Sing Karaoke Songs Unlimited with StarMaker : 26227
Ringtones for iPhone & Ringtone Maker : 25403
Musi - Unlimited Music For YouTube : 25193
AutoRap by Smule : 18202
Spinrilla - Mixtapes For Free : 15053
Napster - Top Music & Radio : 14268
edjing Mix:DJ turntable to remix and scratch music : 13580
Free Music - MP3 Streamer & Playlist Manager Pro : 13443
Free Piano app by Yokee : 13016
Google Play Music : 10118
Certified Mixtapes - Hip Hop Albums & Mixtapes : 9975
TIDAL : 7398


In [90]:
print('~ IOS REFERENCE ~')

for app in ios_final:
    if app[-5] == 'Reference': 
        print(app[1], ':', app[5])
        

~ IOS REFERENCE ~
Bible : 985920
Dictionary.com Dictionary & Thesaurus : 200047
Dictionary.com Dictionary & Thesaurus for iPad : 54175
Google Translate : 26786
Muslim Pro: Ramadan 2017 Prayer Times, Azan, Quran : 18418
New Furniture Mods - Pocket Wiki & Game Tools for Minecraft PC Edition : 17588
Merriam-Webster Dictionary : 16849
Night Sky : 12122
City Maps for Minecraft PE - The Best Maps for Minecraft Pocket Edition (MCPE) : 8535
LUCKY BLOCK MOD ™ for Minecraft PC Edition - The Best Pocket Wiki & Mods Installer Tools : 4693
GUNS MODS for Minecraft PC Edition - Mods Tools : 1497
Guides for Pokémon GO - Pokemon GO News and Cheats : 826
WWDC : 762
Horror Maps for Minecraft PE - Download The Scariest Maps for Minecraft Pocket Edition (MCPE) Free : 718
VPN Express : 14
Real Bike Traffic Rider Virtual Reality Glasses : 8
教えて!goo : 0
Jishokun-Japanese English Dictionary & Translator : 0


Upon review, navigation apps have the highest number of user reviews, but this figure is heavily influenced by Waze and Google Maps, which have close to half a million user reviews together. This also applies to Social Network giants like Facebook, Pinterest, Skype. 

Our aim is to find popular genres, but navigation, social networking or music apps might seem more popular than they really are. The average number of ratings seem to be skewed by very few apps which have hundreds of thousands of user ratings, while the other apps may struggle to get past the 10,000 threshold. We could get a better picture by removing these extremely popular apps for each genre and then rework the averages, but we'll leave this level of detail for later.

Reference apps have 74,942 user ratings on average, but it's actually the Bible and Dictionary.com which skew up the average rating for that genre. However, this niche shows some potential, as the market appears oversaturated with for fun apps, which means a practical app could stand out in the sea of fun apps. 

## Most Popular Apps by Genre on GooglePlay

For the GooglePlay market, we have data about the number of installs, so it should be easier to obtain a clearer understanding about genre popularity. 

However, the install numbers do not appear precise enough as most values are open-ended (100+, 1,000+, 5,000+, etc.)

We will need to convert each string to float, meaning we need to remove commas and plus characters, otherwise the conversion to a float will raise an error. 

In [92]:
display_table(android_final,5)

1,000,000+ : 15.726534296028879
100,000+ : 11.552346570397113
10,000,000+ : 10.548285198555957
10,000+ : 10.198555956678701
1,000+ : 8.393501805054152
100+ : 6.915613718411552
5,000,000+ : 6.825361010830325
500,000+ : 5.561823104693141
50,000+ : 4.7721119133574
5,000+ : 4.512635379061372
10+ : 3.5424187725631766
500+ : 3.2490974729241873
50,000,000+ : 2.3014440433213
100,000,000+ : 2.1322202166064983
50+ : 1.917870036101083
5+ : 0.78971119133574
1+ : 0.5076714801444043
500,000,000+ : 0.2707581227436823
1,000,000,000+ : 0.22563176895306858
0+ : 0.04512635379061372
0 : 0.01128158844765343


In [94]:
categories_android = freq_table(android_final, 1)

for category in categories_android:
    total = 0
    len_category = 0
    for app in android_final:
        category_app = app[1]
        if category_app == category:
            n_installs = app[5]
            n_installs = n_installs.replace(',', '')
            n_installs = n_installs.replace('+', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ':', avg_n_installs)

BOOKS_AND_REFERENCE : 8767811.894736841
COMICS : 817657.2727272727
FINANCE : 1387692.475609756
NEWS_AND_MAGAZINES : 9549178.467741935
HOUSE_AND_HOME : 1331540.5616438356
PARENTING : 542603.6206896552
MEDICAL : 120550.61980830671
ART_AND_DESIGN : 1986335.0877192982
HEALTH_AND_FITNESS : 4188821.9853479853
TRAVEL_AND_LOCAL : 13984077.710144928
PHOTOGRAPHY : 17840110.40229885
LIFESTYLE : 1437816.2687861272
SOCIAL : 23253652.127118643
LIBRARIES_AND_DEMO : 638503.734939759
FOOD_AND_DRINK : 1924897.7363636363
WEATHER : 5074486.197183099
COMMUNICATION : 38456119.167247385
EVENTS : 253542.22222222222
DATING : 854028.8303030303
VIDEO_PLAYERS : 24727872.452830188
ENTERTAINMENT : 11640705.88235294
SHOPPING : 7036877.311557789
FAMILY : 3695641.8198090694
BEAUTY : 513151.88679245283
GAME : 15588015.603248259
SPORTS : 3638640.1428571427
TOOLS : 10801391.298666667
PRODUCTIVITY : 16787331.344927534
EDUCATION : 1833495.145631068
MAPS_AND_NAVIGATION : 4056941.7741935486
PERSONALIZATION : 5201482.61224489

In [96]:
print('~ ANDROID COMMUNICATION ~')

for app in android_final:
    if app[1] == 'COMMUNICATION' and (app[5] == '1,000,000,000+'
                                      or app[5] == '500,000,000+'
                                      or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

~ ANDROID COMMUNICATION ~
WhatsApp Messenger : 1,000,000,000+
imo beta free calls and text : 100,000,000+
Android Messages : 100,000,000+
Google Duo - High Quality Video Calls : 500,000,000+
Messenger – Text and Video Chat for Free : 1,000,000,000+
imo free video calls and chat : 500,000,000+
Skype - free IM & video calls : 1,000,000,000+
Who : 100,000,000+
GO SMS Pro - Messenger, Free Themes, Emoji : 100,000,000+
LINE: Free Calls & Messages : 500,000,000+
Google Chrome: Fast & Secure : 1,000,000,000+
Firefox Browser fast & private : 100,000,000+
UC Browser - Fast Download Private & Secure : 500,000,000+
Gmail : 1,000,000,000+
Hangouts : 1,000,000,000+
Messenger Lite: Free Calls & Messages : 100,000,000+
Kik : 100,000,000+
KakaoTalk: Free Calls & Text : 100,000,000+
Opera Mini - fast web browser : 100,000,000+
Opera Browser: Fast and Secure : 100,000,000+
Telegram : 100,000,000+
Truecaller: Caller ID, SMS spam blocking & Dialer : 100,000,000+
UC Browser Mini -Tiny Fast Private & Secure

In [99]:
print('~ ANDROID: REMOVING COMMUNICATION APPS W/ 100M+ ~')
#removing apps with 100m+

under_100_m = []

for app in android_final:
    n_installs = app[5]
    n_installs = n_installs.replace(',', '')
    n_installs = n_installs.replace('+', '')
    if (app[1] == 'COMMUNICATION') and (float(n_installs) < 100000000):
        under_100_m.append(float(n_installs))
        
sum(under_100_m) / len(under_100_m)

~ ANDROID: REMOVING COMMUNICATION APPS W/ 100M+ ~


3603485.3884615386

In [101]:
print('~ ANDROID BOOKS & REFERENCE ~')

for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE':
        print(app[0], ':', app[5])

~ ANDROID BOOKS & REFERENCE ~
E-Book Read - Read Book for free : 50,000+
Download free book with green book : 100,000+
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Free Panda Radio Music : 100,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
English Grammar Complete Handbook : 500,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
Google Play Books : 1,000,000,000+
AlReader -any text book reader : 5,000,000+
Offline English Dictionary : 100,000+
Offline: English to Tagalog Dictionary : 500,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
Recipes of Prophetic Medicine for free : 500,000+
ReadEra – free ebook reader : 1,000,000+
Anonymous caller detection : 10,000+
Ebook Reader : 5,000,000+
Litnet - E-books : 100,000+
Read books online : 5,000,000+
English to Urdu Dictionary : 500,000+
eBoox: book reader fb2 epub zip : 1,000,000+
English Persian Dictionary : 500,000+
Flybook : 500,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000

In [103]:
print('~ ANDROID: POPULAR BOOKS & REFERENCE APPS ~')
for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000,000+'
                                            or app[5] == '500,000,000+'
                                            or app[5] == '100,000,000+'):
        print(app[0], ':', app[5])

~ ANDROID: POPULAR BOOKS & REFERENCE APPS ~
Google Play Books : 1,000,000,000+
Bible : 100,000,000+
Amazon Kindle : 100,000,000+
Wattpad 📖 Free Books : 100,000,000+
Audiobooks from Audible : 100,000,000+


In [104]:
print('~ ANDROID: MID BOOKS & REFERENCE APPS')

for app in android_final:
    if app[1] == 'BOOKS_AND_REFERENCE' and (app[5] == '1,000,000+'
                                            or app[5] == '5,000,000+'
                                            or app[5] == '10,000,000+'
                                            or app[5] == '50,000,000+'):
        print(app[0], ':', app[5])

~ ANDROID: MID BOOKS & REFERENCE APPS
Wikipedia : 10,000,000+
Cool Reader : 10,000,000+
Book store : 1,000,000+
FBReader: Favorite Book Reader : 10,000,000+
Free Books - Spirit Fanfiction and Stories : 1,000,000+
AlReader -any text book reader : 5,000,000+
FamilySearch Tree : 1,000,000+
Cloud of Books : 1,000,000+
ReadEra – free ebook reader : 1,000,000+
Ebook Reader : 5,000,000+
Read books online : 5,000,000+
eBoox: book reader fb2 epub zip : 1,000,000+
All Maths Formulas : 1,000,000+
Ancestry : 5,000,000+
HTC Help : 10,000,000+
Moon+ Reader : 10,000,000+
English-Myanmar Dictionary : 1,000,000+
Golden Dictionary (EN-AR) : 1,000,000+
All Language Translator Free : 1,000,000+
Aldiko Book Reader : 10,000,000+
Dictionary - WordWeb : 5,000,000+
50000 Free eBooks & Free AudioBooks : 5,000,000+
Al-Quran (Free) : 10,000,000+
Al Quran Indonesia : 10,000,000+
Al'Quran Bahasa Indonesia : 10,000,000+
Al Quran Al karim : 1,000,000+
Al Quran : EAlim - Translations & MP3 Offline : 5,000,000+
Koran R

It appears creating an app that will fit into the Book / Reference genre would be beneficial for both GooglePlay and App Store Markets.

However, the market is saturated with libraries, so possibly take a recent popular book and adding special features such a daily quotes from the book, audtio version of the book, quizzes, forum, dictionary imbedd in the app to prevent the user from hai
However, it looks like the market is already full of libraries, so we need to add some special features besides the raw version of the book. This might include daily quotes from the book, an audio version of the book, quizzes on the book, a forum where people can discuss the book, etc.